physicalplan/step_invariant/step_invariant.go

```diff
@@ -0,0 +1,114 @@
+// Copyright (c) The Thanos Community Authors.
+// Licensed under the Apache License 2.0.
+
+package step_invariant
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/efficientgo/core/errors"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/promql/parser"
+
+	"github.com/thanos-community/promql-engine/physicalplan/model"
+)
+
+type stepInvariantOperator struct {
+	vectorPool       *model.VectorPool
+	next             model.VectorOperator
+	once             sync.Once
+	duplicateResults bool
+
+	series []labels.Labels
+
+	mint int64
+	maxt int64
+	step int64
+}
+
+func (u *stepInvariantOperator) Explain() (me string, next []model.VectorOperator) {
+	return "[*stepInvariantOperator]", []model.VectorOperator{u.next}
+}
+
+func NewStepInvariantOperator(
+	pool *model.VectorPool,
+	next model.VectorOperator,
+	expr parser.Expr,
+	mint, maxt time.Time,
+	step time.Duration,
+) (model.VectorOperator, error) {
+	interval := step.Milliseconds()
+	// We set interval to be at least 1.
+	if interval == 0 {
+		interval = 1
+	}
+	u := &stepInvariantOperator{
+		vectorPool:       pool,
+		next:             next,
+		mint:             mint.UnixMilli(),
+		maxt:             maxt.UnixMilli(),
+		step:             interval,
+		duplicateResults: true,
+	}
+	// We do not duplicate results for range selectors since result is a matrix
+	// with their unique timestamps which does not depend on the step.
+	switch expr.(type) {
+	case *parser.MatrixSelector, *parser.SubqueryExpr:
+		u.duplicateResults = false
+	}
+
+	return u, nil
+}
+
+func (u *stepInvariantOperator) Series(ctx context.Context) ([]labels.Labels, error) {
+	var err error
+	u.once.Do(func() {
+		u.series, err = u.next.Series(ctx)
+	})
+	if err != nil {
+		return nil, err
+	}
+	return u.series, nil
+}
+
+func (u *stepInvariantOperator) GetPool() *model.VectorPool {
+	return u.next.GetPool()
+}
+
+func (u *stepInvariantOperator) Next(ctx context.Context) ([]model.StepVector, error) {
+	in, err := u.next.Next(ctx)
+	if err != nil {
+		return nil, err
+	}
+	if in == nil || len(in) == 0 || len(in[0].Samples) == 0 {
+		return nil, nil
+	}
+	if !u.duplicateResults {
```

Comment:
What does this flag indicate?

Reply:
Used in https://github.com/prometheus/prometheus/blob/main/promql/engine.go#L1675.
If not matrix or subquery then we will duplicate the values with steps.